// hover-tooltip/dist/Tooltip-apklNNHV.js
import React, { useState, useRef, useEffect } from 'react';
import { AnimatePresence, motion } from 'framer-motion';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */


var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var Tooltip = function (_a) {
    var children = _a.children, text = _a.text, _b = _a.position, position = _b === void 0 ? "top" : _b, _c = _a.theme, theme = _c === void 0 ? "modern" : _c, _d = _a.delay, delay = _d === void 0 ? 0.1 : _d, _e = _a.arrow, arrow = _e === void 0 ? true : _e, _f = _a.maxWidth, maxWidth = _f === void 0 ? "250px" : _f, _g = _a.className, className = _g === void 0 ? "" : _g, _h = _a.animation, animation = _h === void 0 ? "smooth" : _h, _j = _a.duration, duration = _j === void 0 ? 0.3 : _j, _k = _a.distance, distance = _k === void 0 ? 8 : _k, _l = _a.showOnClick, showOnClick = _l === void 0 ? false : _l, _m = _a.persistent, persistent = _m === void 0 ? false : _m, _o = _a.fontSize, fontSize = _o === void 0 ? "sm" : _o, _p = _a.rich, rich = _p === void 0 ? false : _p;
    var _q = useState(false), visible = _q[0], setVisible = _q[1];
    var tooltipRef = useRef(null);
    var triggerRef = useRef(null);
    useEffect(function () {
        var handleClickOutside = function (event) {
            if (persistent &&
                tooltipRef.current &&
                triggerRef.current &&
                !tooltipRef.current.contains(event.target) &&
                !triggerRef.current.contains(event.target)) {
                setVisible(false);
            }
        };
        document.addEventListener("mousedown", handleClickOutside);
        return function () { return document.removeEventListener("mousedown", handleClickOutside); };
    }, [persistent]);
    var getAnimationVariants = function () {
        var baseOpacity = { opacity: 0 };
        var animations = {
            smooth: {
                hidden: __assign(__assign(__assign(__assign(__assign({}, baseOpacity), (position === "top" && { y: distance, x: "-50%" })), (position === "bottom" && { y: -distance, x: "-50%" })), (position === "left" && { x: distance, y: "-50%" })), (position === "right" && { x: -distance, y: "-50%" })),
                visible: {
                    opacity: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "spring",
                        damping: 25,
                        stiffness: 300,
                    },
                },
            },
            pop: {
                hidden: __assign(__assign(__assign(__assign(__assign(__assign({}, baseOpacity), { scale: 0.9 }), (position === "top" && { y: distance, x: "-50%" })), (position === "bottom" && { y: -distance, x: "-50%" })), (position === "left" && { x: distance, y: "-50%" })), (position === "right" && { x: -distance, y: "-50%" })),
                visible: {
                    opacity: 1,
                    scale: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "spring",
                        damping: 15,
                        stiffness: 400,
                    },
                },
            },
            fade: {
                hidden: {
                    opacity: 0,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                },
                visible: {
                    opacity: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "tween",
                        ease: "easeInOut",
                    },
                },
            },
            scale: {
                hidden: {
                    opacity: 0,
                    scale: 0.5,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                },
                visible: {
                    opacity: 1,
                    scale: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "spring",
                        damping: 20,
                        stiffness: 350,
                    },
                },
            },
            elastic: {
                hidden: __assign(__assign(__assign(__assign(__assign(__assign({}, baseOpacity), { scale: 0.8 }), (position === "top" && { y: distance * 1.5, x: "-50%" })), (position === "bottom" && { y: -distance * 1.5, x: "-50%" })), (position === "left" && { x: distance * 1.5, y: "-50%" })), (position === "right" && { x: -distance * 1.5, y: "-50%" })),
                visible: {
                    opacity: 1,
                    scale: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "spring",
                        damping: 8,
                        stiffness: 300,
                        mass: 0.8,
                    },
                },
            },
            shift: {
                hidden: __assign(__assign(__assign(__assign(__assign({}, baseOpacity), (position === "top" && { y: distance, x: "-50%" })), (position === "bottom" && { y: -distance, x: "-50%" })), (position === "left" && { x: distance, y: "-50%" })), (position === "right" && { x: -distance, y: "-50%" })),
                visible: {
                    opacity: 1,
                    y: position === "top" || position === "bottom" ? 0 : "-50%",
                    x: position === "left" || position === "right" ? 0 : "-50%",
                    transition: {
                        type: "spring",
                        damping: 20,
                        stiffness: 300,
                    },
                },
            },
        };
        return animations[animation];
    };
    var getThemeClasses = function () {
        var themes = {
            modern: "bg-slate-900 text-white shadow-lg shadow-slate-900/20",
            elegant: "bg-white/95 text-slate-900 shadow-xl border border-slate-200/50 backdrop-blur-sm",
            frost: "bg-white/20 backdrop-blur-md text-white shadow-lg border border-white/30",
            neon: "bg-black/90 text-white shadow-lg border-2 border-purple-500/50 shadow-purple-500/20",
            soft: "bg-slate-100 text-slate-700 shadow-md border border-slate-200",
            glass: "bg-white/10 backdrop-blur-lg border border-white/20 text-white shadow-xl",
        };
        return themes[theme];
    };
    var getFontSizeClass = function () {
        var sizes = {
            sm: "text-sm",
            base: "text-base",
            lg: "text-lg",
        };
        return sizes[fontSize];
    };
    var getPositionClasses = function () {
        var positions = {
            top: "bottom-full left-1/2 mb-2",
            bottom: "top-full left-1/2 mt-2",
            left: "right-full top-1/2 mr-2",
            right: "left-full top-1/2 ml-2",
        };
        return positions[position];
    };
    var getArrowClasses = function () {
        if (!arrow)
            return "";
        var baseClasses = "after:content-[''] after:absolute ";
        var positions = {
            top: "after:top-full after:left-1/2 after:-translate-x-1/2 after:border-[6px] after:border-t-current after:border-x-transparent after:border-b-transparent",
            bottom: "after:bottom-full after:left-1/2 after:-translate-x-1/2 after:border-[6px] after:border-b-current after:border-x-transparent after:border-t-transparent",
            left: "after:left-full after:top-1/2 after:-translate-y-1/2 after:border-[6px] after:border-l-current after:border-y-transparent after:border-r-transparent",
            right: "after:right-full after:top-1/2 after:-translate-y-1/2 after:border-[6px] after:border-r-current after:border-y-transparent after:border-l-transparent",
        };
        var colors = {
            modern: "after:border-t-slate-900 after:border-b-slate-900 after:border-l-slate-900 after:border-r-slate-900",
            elegant: "after:border-t-white/95 after:border-b-white/95 after:border-l-white/95 after:border-r-white/95",
            frost: "after:border-t-white/20 after:border-b-white/20 after:border-l-white/20 after:border-r-white/20",
            neon: "after:border-t-black/90 after:border-b-black/90 after:border-l-black/90 after:border-r-black/90",
            soft: "after:border-t-slate-100 after:border-b-slate-100 after:border-l-slate-100 after:border-r-slate-100",
            glass: "after:border-t-white/10 after:border-b-white/10 after:border-l-white/10 after:border-r-white/10",
        };
        return "".concat(baseClasses, " ").concat(positions[position], " ").concat(colors[theme]);
    };
    var handleMouseEvents = function (event) {
        if (showOnClick && event.type === "click") {
            setVisible(!visible);
        }
        else if (!showOnClick) {
            setVisible(event.type === "mouseenter");
        }
    };
    return (React.createElement("div", { className: "relative inline-block" },
        React.createElement("div", { ref: triggerRef, onClick: handleMouseEvents, onMouseEnter: handleMouseEvents, onMouseLeave: function () { return !persistent && setVisible(false); }, className: "cursor-pointer" }, children),
        React.createElement(AnimatePresence, null, visible && (React.createElement(motion.div, { ref: tooltipRef, initial: "hidden", animate: "visible", exit: "hidden", variants: getAnimationVariants(), transition: {
                duration: duration,
                delay: visible ? delay : 0,
            }, style: { maxWidth: maxWidth }, className: "\n              fixed px-4 py-2 rounded-lg\n              font-medium tracking-wide\n              ".concat(getFontSizeClass(), "\n              ").concat(getThemeClasses(), "\n              ").concat(getPositionClasses(), "\n              ").concat(getArrowClasses(), "\n              ").concat(className, "\n            ") }, rich ? React.createElement("div", { dangerouslySetInnerHTML: { __html: text } }) : text)))));
};

export { Tooltip as T };
